### ISSUE TITLE
Merge with Apache Thrift

### ISSUE THREAD

**Issue by daixtrose** [Microsoft/thrifty#32]
I am deeply concerned about the road that Thrift took during the past years:
- Facebook donated Thrift to the Apache Foundation
- Facebook forked its own [incompatible branch](https://github.com/facebook/fbthrift) for Linux only, adding new features like `float` support
- Microsoft [optimized the Apache Thrift JAVA code](/Microsoft/thrifty)

What I fear is that we end up with 4 or 5 different versions of one of the best IPC libraries  of the world and that Thrift wil simply die, die, die due to this.

Could you please explain what hinders you to merge Thrifty back into [Apache Thrift ](https://thrift.apache.org/)? Is the Apache developer team not responsive enough? Are there technical issues? Are there incompatibilities?

**Comment by benjamin-bader** [comment]
Hi @daixtrose,

**TL;DR:**
- Because we took semantic shortcuts, we will likely remain an independent implementation
- Thrifty is *not* general-purpose Thrift, and only makes sense in an Android application
- Thrifty is binary-compatible with the Apache implementation, and we test every build for it
- Any client->server incompatibility *is a bug, and please file an issue for it*

The long version:

These are definitely valid concerns.  Our goal with this project was to radically reduce the code size both of the runtime and of generated code; we made the gains we did because we started from a few subtly different assumptions than the Apache implementation.  The thorniest difference turns out to be the treatment of `required`, `optional`, and the default requiredness semantics.  If for no other reason than this, it's not so simple to reconcile Thrifty with Apache Thrift.

Further, the runtime assumptions made by Apache's implementation (e.g. that field-setness is tracked independently of field value) are violated here.  This works because a) Thrift has no concept of `null`, so that's available to us to use, and b) this is explicitly a client-only library, and we can take certain semantic shortcuts that wouldn't hold on the server.  The Thrift maintainers have affirmed this particular design point several times already, so there's that. 

I can't really speak to the responsiveness of the Thrift developer team, simply because we haven't really engaged with them.  The possibility of merging Thrifty into Apache never really entered our minds, honestly, because it seems fundamentally a very difficult proposition.  I don't believe that we would be opposed, for the record.

Finally, I would like to note that Thrifty doesn't make sense outside of Android.  There's no real reason to use an "unofficial" Thrift implementation unless, as is the case on Android, platform constraints make the Apache version unpalatable.  Bearing that in mind, *the* major criteria for Thrifty is that it is binary-compatible with Apache.  It doesn't matter how svelte the generated code is - if we aren't compatible, *we have failed*.  To that end, we test for compatibility on every build.  **If you find a deviation, it is a bug, so please report it**.

Not quite the answer you were looking for, but I hope that I have shed at least some light here.

**Comment by daixtrose** [comment]
I can follow your arguments. Yet I still think that if _Thrifty_ is a binary-compatible client (only) implementation it still makes sense to incorporate it into the Apache Thrift source tree.

- many language specific implementatons have their flaws and/or limitations (e.g. the [C# implementation has some problem with ipv4/ipv6 dual mode](https://issues.apache.org/jira/browse/THRIFT-3000)), but that is OK
- keeping things together might create the level of awareness neccesary for further development

**Comment by benjamin-bader** [comment]
Sorry to have been unclear.  While we are not opposed in principle, there is only small likelihood of merging Thrifty with Apache.  Right now, none of us has the time to pursue such a merge.

It should be made clear that Thrifty is *not* intended to compete with or supplant Apache - it fills a very specific niche that, currently, Apache does not.

### Regarding binary compatibility and client-only:

When I say *binary-compatible*, I refer to serialized data on the wire *in the use case of a Thrifty client communicating with a server*.  This is true of all current use of Thrifty of which I'm aware.  I exclude the use of Thrifty as a server platform, because 1) the runtime does not contain such support, and b) the limitations I alluded to come in to play in generated service handlers.

`required` and `optional` are officially underspecified (see the giant comment in `Requiredness.java` for details), and we take advantage of that: in Thrifty, `null` equals `unset`.  This is, strictly speaking, contrary to the stated understanding of the Apache devs, where a default field can be both `null` and `set`.  While this state doesn't make much sense in Java, it is certainly legal and quite common, apparently.  This state cannot exist in Thrifty, but spec-compliant Thrift services must be prepared to handle it.

Thankfully, from what I can tell services will never return `null` for a `required` field, and so the problem is avoided entirely - so long as we are client-only.

I cannot see a way to making Thrifty work server-side without breaking this fundamental design of `null` == `unset`, and if we break that then we lose basically everything we've gained: back come the `isSet`, `unset`, `setIsSet`, and `validate` methods, one for each generated struct.  Given that Thrifty's stated purpose is to reduce the number of generated methods, I'm strongly disinclined to take this path.

There isn't really any benefit to doing so, anyways - Thrifty's benefit is that the generated code is 20% the size of Apache's, but code size is really only a constraint on Android.  On the server, you can have as many methods as you want!  No need to burden oneself with extra work when there is a perfectly good, well-tested, server-capable implementation from Apache.
